Neos.ContentRepository/Classes/Domain/Repository/NodeDataRepository.php

```diff
@@ -1238,10 +1238,11 @@ protected function addDimensionJoinConstraintsToQueryBuilder(QueryBuilder $query
      * @param array $nodes NodeData result with multiple and duplicate identifiers (different nodes and redundant results for node variants with different dimensions)
      * @param array $workspaces
      * @param array $dimensions
+     * @param boolean $preferNonShadowNodes if set to TRUE and we have two nodes with the same dimension positions, we prefer the non-shadow variant of it.
      * @return array Array of unique node results indexed by identifier
      * @throws Exception\NodeException
      */
-    protected function reduceNodeVariantsByWorkspacesAndDimensions(array $nodes, array $workspaces, array $dimensions)
+    protected function reduceNodeVariantsByWorkspacesAndDimensions(array $nodes, array $workspaces, array $dimensions, $preferNonShadowNodes = false)
```

Comment:
I know, it's a complex beast which isn't easy to tame, but that new parameter smells like one of those things you'll stumble over in a couple of months and wonder why the heck this is configurable or which implications it has to pass true or false.

If that is in option for you, I suggest that you somehow refactor this behavior in a more speaking method call.